scipy/stats/_distn_infrastructure.py

```diff
@@ -3836,3 +3884,423 @@ def get_distribution_names(namespace_pairs, rv_base_class):
+                   f"`{name}` parameter.")
+        raise ValueError(message)
+
+    if user_bounds[0] > user_bounds[1]:
+        message = (f"There are no values for `{name}` on the interval "
+                   f"{list(user_bounds)}.")
+        raise ValueError(message)
+
+    bounds = (max(user_bounds[0], shape_domain[0]),
+              min(user_bounds[1], shape_domain[1]))
+
+    if integral and (np.ceil(bounds[0]) > np.floor(bounds[1])):
+        message = (f"There are no integer values for `{name}` on the interval "
+                   f"defined by the user-provided bounds and the domain "
+                   "of the distribution.")
+        raise ValueError(message)
+    elif not integral and (bounds[0] > bounds[1]):
+        message = (f"There are no values for `{name}` on the interval "
+                   f"defined by the user-provided bounds and the domain "
+                   "of the distribution.")
+        raise ValueError(message)
+
+    if not np.all(np.isfinite(bounds)):
+        message = (f"The intersection of user-provided bounds for `{name}` "
+                   f"and the domain of the distribution is not finite. Please "
+                   f"provide finite bounds for shape `{name}` in "
+                   f"`{bounds_type}`.")
+        raise ValueError(message)
+
+    return bounds
+
+
+class FitResult:
+    r"""Result of fitting a discrete or continuous distribution to data
+
+    Attributes
+    ----------
+    dist : scipy.stats.rv_continuous or scipy.stats.rv_discrete
+        The distribution of which parameters were fit.
+    data : 1D array_like
+        The data to which the distribution was fit.
+    nllf : callable
+        The negative log-likehood function for the given `data`. Accepts a
+        single tuple containing the shapes, location, and scale of the
+        distribution.
+    params : namedtuple
+        A namedtuple containing the maximum likelihood estimates of the
+        shape parameters, location, and (if applicable) scale of the
+        distribution.
+    objective_val : float
+        The value of the negative log likelihood function evaluated with
+        the fitted shapes, i.e. ``result.nllf(result.params)``.
+    success : bool or None
+        Whether the optimizer considered the optimization to terminate
+        successfully or not.
+    message : str or None
+        Any status message provided by the optimizer.
+
+    """
+
+    def __init__(self, dist, data, discrete, nllf, res):
+        self.dist = dist
+        self.data = data
+        self.discrete = discrete
+        self.pxf = getattr(dist, "pmf", None) or getattr(dist, "pdf", None)
+        self.success = getattr(res, "success", None)
+        self.message = getattr(res, "message", None)
+
+        shape_names = [] if dist.shapes is None else dist.shapes.split(", ")
+        if getattr(dist, "pdf", False):
+            FitShapes = namedtuple('FitShapes', shape_names + ['loc', 'scale'])
```

Comment:
I thought it would be nice for the shapes to be in a namedtuple so that it's very explicit which parameter is which.
We might want to call this `FitParams`, though, as I'm pretty careful elsewhere to make a distinction between _shape_ parameters and all parameters (including `loc` and `scale`).